code/ui/ui_main.c

```diff
@@ -263,19 +265,39 @@ void UI_WideScreenMode(qboolean on) {
 UI_UpdateWidescreen
 =================
 */
-extern vmCvar_t ui_widescreen;
 static void UI_UpdateWidescreen(void) {
 	if (ui_widescreen.integer && mvapi >= 3) {
-		uiInfo.screenWidth = (float)SCREEN_HEIGHT * uiInfo.uiDC.glconfig.vidWidth / uiInfo.uiDC.glconfig.vidHeight;
-	}
-	else {
+		if ( uiInfo.uiDC.glconfig.vidWidth >= uiInfo.uiDC.glconfig.vidHeight ) {
+			uiInfo.screenWidth = (float)SCREEN_HEIGHT * uiInfo.uiDC.glconfig.vidWidth / uiInfo.uiDC.glconfig.vidHeight;
+			uiInfo.screenHeight = (float)SCREEN_HEIGHT;
+			uiInfo.portraitMode = qfalse;
+		} else {
+			uiInfo.screenWidth = (float)SCREEN_WIDTH;
+			uiInfo.screenHeight = (float)SCREEN_WIDTH * uiInfo.uiDC.glconfig.vidHeight / uiInfo.uiDC.glconfig.vidWidth;
+			uiInfo.portraitMode = qtrue;
+		}
+	} else {
 		uiInfo.screenWidth = (float)SCREEN_WIDTH;
+		uiInfo.screenHeight = (float)SCREEN_HEIGHT;
+		uiInfo.portraitMode = qfalse;
 	}
+
 	uiInfo.screenXFactor = (float)SCREEN_WIDTH / uiInfo.screenWidth;
 	uiInfo.screenXFactorInv = uiInfo.screenWidth / (float)SCREEN_WIDTH;
 
-	if (mvapi >= 3 && ui_widescreen.integer != 2)
-		trap_MVAPI_SetVirtualScreen(uiInfo.screenWidth, (float)SCREEN_HEIGHT);
+	uiInfo.screenYFactor = (float)SCREEN_HEIGHT / uiInfo.screenHeight;
+	uiInfo.screenYFactorInv = uiInfo.screenHeight / (float)SCREEN_HEIGHT;
+
+	if (uiInfo.portraitMode && isMainMenu) {
+		uiInfo.screenWidth = (float)SCREEN_WIDTH;
```

Comment:
uiInfo.screenXFactor is a cached result of uiInfo.screenWidth / uiInfo.screenHeight and nothing more. it must be always in sync so that implementing widescreen-correct drawing is straightforward. There can be no assumptions made on how it is used, it's just a shorthand.